kubeproxy/controllers/kubeproxy_controller.go

```diff
@@ -0,0 +1,216 @@
+	if err := o.MutateContainers(func(container map[string]interface{}) error {
+		envList, _, err := unstructured.NestedSlice(container, "env")
+		if err != nil {
+			return fmt.Errorf("error reading container env: %v", err)
+		}
+		for k, v := range env {
+			foundK := false
+			for _, e := range envList {
+				m, ok := e.(map[string]interface{})
+				if !ok {
+					return fmt.Errorf("env var was not an object: %v", err)
+				}
+				name, found, err := unstructured.NestedString(m, "name")
+				if err != nil {
+					return err
+				}
+				if found && name == k {
+					if err := unstructured.SetNestedField(m, v, "value"); err != nil {
+						return err
+					}
+					foundK = true
+				}
+			}
+			if !foundK {
+				envList = append(envList, map[string]interface{}{
+					"name":  k,
+					"value": v,
+				})
+			}
+		}
+		// Sort env values by name so we have a consistent order
+		sort.Slice(envList, func(i, j int) bool {
+			mapI, okI := envList[i].(map[string]interface{})
+			mapJ, okJ := envList[j].(map[string]interface{})
+			if !okJ {
+				return false
+			}
+			if !okI {
+				return true
+			}
+			kI, foundI := mapI["name"]
+			kJ, foundJ := mapJ["name"]
+			if !foundJ {
+				return false
+			}
+			if !foundI {
+				return true
+			}
+			return kI.(string) < kJ.(string)
+		})
+		if err := unstructured.SetNestedSlice(container, envList, "env"); err != nil {
+			return fmt.Errorf("error setting env vars: %v", err)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// OverrideApiserver sets environment variables, hostNetwork & dns policy so that we don't rely on kube-proxy to reach the api server
+func OverrideApiserver(ctx context.Context, o declarative.DeclarativeObject, manifest *manifest.Objects) error {
+	for _, o := range manifest.Items {
+		if o.Kind == "DaemonSet" {
+			// KubeProxy (and kubelet) are special: it has to find the apiserver directly,
+			// other clients use the VIP that kubeproxy configures.
+			//
+			// Because of this we need a special env var to allow for node kube-proxies to have a different endpoint than the operator
+			// (If the operator runs on the master, it can use 127.0.0.1; we definitely can't use that for the node kube-proxy)
+			// We still fall-back to the existing mechanisms - KUBERNETES_SERVICE_HOST, then a hard-coded default value
+			master := os.Getenv("KUBEPROXY_KUBERNETES_SERVICE_HOST")
+			if master == "" {
+				master = os.Getenv("KUBERNETES_SERVICE_HOST")
+			}
+			if master == "" {
+				master = "kubernetes-master"
+				klog.Warningf("using fallback for KUBERNETES_SERVICE_HOST: %v", master)
+			}
+			port := os.Getenv("KUBERNETES_SERVICE_PORT")
+			if port == "" {
+				port = "443"
+			}
+			env := map[string]string{"KUBERNETES_SERVICE_HOST": master, "KUBERNETES_SERVICE_PORT": port}
+			if err := SetEnvironmentVariables(o, env); err != nil {
```

Comment:
I don't think that this is the best way for a long term solution. I would recommend having a field in `KubeProxySpec` that explicitly sets that value. But it's fine to use this approach as long as the value in the CRD is not set.